File: haskell/plugins/Onepassword.hs
-- | 1Password Connect, as a plugin.
--
-- A port of typescript/plugins/onepassword.ts, which is canonical.

module Onepassword
  ( onepassword,
    onepasswordprovider,
  )
where

import Control.Monad (when)
import Data.IORef (newIORef, readIORef, writeIORef)
import Data.Maybe (fromMaybe)
import Defs (Definition)
import Httpjson (Answer (..), fetchjson, uriescape)
import qualified Json
import Names (checkname)
import Provider (Provider (..), forced)
import Providers
  ( ProviderSpec (..),
    checkaddr,
    fail',
    providerplugin,
    trimslash,
  )


-- | 1Password, through a Connect server.
--
-- The item titled @api.token@ (titles keep their dots), in the named
-- vault. The value is the field with purpose PASSWORD, or the field
-- labelled @value@. A vault that cannot be found is an error - config
-- names it, so its absence is a broken store, not a missing secret.
onepasswordprovider :: ProviderSpec -> IO Provider
onepasswordprovider spec = do
  vaultid <- newIORef Nothing

  let auth = [("authorization", "Bearer " ++ spectoken spec)]

      resolvevault useaddr = do
        let want = specvault spec
        when (null want) $ fail' "sekreto: onepassword: no vault"

        res <- fetchjson "GET" (useaddr ++ "/v1/vaults") auth Nothing

        entries <- case Json.asarr (ansbody res) of
          Just found | 200 == ansstatus res -> pure found
          _ -> fail' ("sekreto: onepassword error: " ++ show (ansstatus res) ++ ": listing vaults")

        let matches entry =
              Just want == Json.text (Json.dig (Just entry) ["id"])
                || Just want == Json.text (Json.dig (Just entry) ["name"])

        case filter matches entries of
          (entry : _) -> pure (fromMaybe "" (Json.text (Json.dig (Just entry) ["id"])))
          [] -> fail' ("sekreto: onepassword: no vault named " ++ want)

  pure
    Provider
      { lookupsecret = \name -> do
          _ <- forced (checkname name)

          let useaddr = trimslash (specaddr spec)
          when (null useaddr) $ fail' "sekreto: onepassword: no addr"
          checkaddr useaddr

          held <- readIORef vaultid
          vid <- case held of
            Just found -> pure found
            Nothing -> do
              found <- resolvevault useaddr
              writeIORef vaultid (Just found)
              pure found

          let filterexp = uriescape ("title eq \"" ++ name ++ "\"")
              findurl = useaddr ++ "/v1/vaults/" ++ vid ++ "/items?filter=" ++ filterexp

          found <- fetchjson "GET" findurl auth Nothing

          items <- case Json.asarr (ansbody found) of
            Just entries | 200 == ansstatus found -> pure entries
            _ ->
              fail'
                ("sekreto: onepassword error: " ++ show (ansstatus found) ++ ": finding " ++ name)

          case items of
            [] -> pure Nothing
            (entry : _) -> do
              let itemid = fromMaybe "" (Json.text (Json.dig (Just entry) ["id"]))
                  readurl = useaddr ++ "/v1/vaults/" ++ vid ++ "/items/" ++ itemid

              item <- fetchjson "GET" readurl auth Nothing

              when (200 /= ansstatus item) $
                fail'
                  ( "sekreto: onepassword error: "
                      ++ show (ansstatus item)
                      ++ ": reading "
                      ++ name
                  )

              let fields = fromMaybe [] (Json.asarr (Json.dig (ansbody item) ["fields"]))

                  -- The FIRST field carrying this role, if any. Deliberately
                  -- NOT "the first field carrying this role whose value is
                  -- text": those are different, and conflating them is a
                  -- wrong-secret bug rather than a cosmetic one.
                  firstwith role wanted =
                    case filter
                      (\field -> Just wanted == Json.asstr (Json.dig (Just field) [role]))
                      fields of
                      (field : _) -> Just field
                      [] -> Nothing

                  valueof field = Json.text (Json.dig (Just field) ["value"])

              -- Two passes, purpose first -- and the FIRST pass is TERMINAL.
              -- Canonical returns from inside the loop
              -- (kotlin/src/Providers.kt:1417 and the typescript plugin), so
              -- a PASSWORD field whose value is null is a MISS and must not
              -- fall through to the label pass. Returning Nothing from the
              -- purpose lookup for both "no such field" and "field with a
              -- null value" made it fall through, and the label pass then
              -- answered with a DIFFERENT field's value: the caller got a
              -- secret where canonical gives it nothing.
              pure
                ( case firstwith "purpose" "PASSWORD" of
                    Just field -> valueof field
                    Nothing -> maybe Nothing valueof (firstwith "label" "value")
                ),
        describe = "onepassword:" ++ specvault spec
      }


-- | @onepassword@, as a voxgig/plugin definition.
onepassword :: Definition
onepassword = providerplugin "onepassword" onepasswordprovider

File: ocaml/plugins/onepassword.ml
(* The `onepassword` provider kind, as a voxgig/plugin definition.

   1Password Connect over HTTPS.

   A port of typescript/plugins/onepassword.ts, which is canonical. *)

open Secret
open Provider
open Httpjson

(* 1Password, through a Connect server.

   The item titled `api.token` (titles KEEP THEIR DOTS), in the named vault.
   The value is the field with purpose PASSWORD, or the field labelled
   `value`. A vault that cannot be found is an ERROR, not a miss - config
   names it, so its absence is a broken store rather than a missing
   secret. *)
let onepassword_provider (addr : string) (token : string) (vault : string) : provider =
  let vaultid = ref None in

  let auth () = [ ("authorization", "Bearer " ^ token) ] in

  let resolvevault useaddr =
    if "" = vault then fail "sekreto: onepassword: no vault";

    let res = fetchjson ~headers:(auth ()) "GET" (useaddr ^ "/v1/vaults") in
    let list = Json.oasarr res.jbody in

    if 200 <> res.status || None = list then
      fail ("sekreto: onepassword error: " ^ string_of_int res.status ^ ": listing vaults");

    let entries = Option.get list in
    let found =
      List.find_opt
        (fun entry ->
          Some vault = Json.otext (Json.dig entry [ "id" ])
          || Some vault = Json.otext (Json.dig entry [ "name" ]))
        entries
    in

    match found with
    | Some entry -> Option.value ~default:"" (Json.otext (Json.dig entry [ "id" ]))
    | None -> fail ("sekreto: onepassword: no vault named " ^ vault)
  in

  {
    lookup =
      (fun name ->
        ignore (checkname name);

        let useaddr = trimslash addr in
        if "" = useaddr then fail "sekreto: onepassword: no addr";
        checkaddr useaddr;

        let id =
          match !vaultid with
          | Some found -> found
          | None ->
            let resolved = resolvevault useaddr in
            vaultid := Some resolved;
            resolved
        in

        let filter = Sigv4.uriescape ("title eq \"" ^ name ^ "\"") in
        let found =
          fetchjson ~headers:(auth ()) "GET" (useaddr ^ "/v1/vaults/" ^ id ^ "/items?filter=" ^ filter)
        in

        let items = Json.oasarr found.jbody in
        if 200 <> found.status || None = items then
          fail ("sekreto: onepassword error: " ^ string_of_int found.status ^ ": finding " ^ name);

        match Option.get items with
        (* No item with that title: a miss, and the chain carries on. *)
        | [] -> None
        | head :: _ ->
          let itemid = Option.value ~default:"" (Json.otext (Json.dig head [ "id" ])) in
          let item =
            fetchjson ~headers:(auth ()) "GET" (useaddr ^ "/v1/vaults/" ^ id ^ "/items/" ^ itemid)
          in

          if 200 <> item.status then
            fail ("sekreto: onepassword error: " ^ string_of_int item.status ^ ": reading " ^ name);

          let fields =
            match Json.oasarr (Json.odig item.jbody [ "fields" ]) with
            | Some list -> list
            | None -> []
          in

          (* Two full passes, in this order. *)
          let bypurpose =
            List.find_opt (fun field -> Some "PASSWORD" = Json.oasstr (Json.dig field [ "purpose" ])) fields
          in

          (match bypurpose with
          | Some field -> Json.otext (Json.dig field [ "value" ])
          | None -> (
            match
              List.find_opt (fun field -> Some "value" = Json.oasstr (Json.dig field [ "label" ])) fields
            with
            | Some field -> Json.otext (Json.dig field [ "value" ])
            | None -> None)));
    describe = (fun () -> "onepassword:" ^ vault);
  }

let plugin () : Defs.definition =
  providerplugin "onepassword" (fun spec ->
      onepassword_provider spec.addr spec.token spec.vault)
